File: src/factories/internal/metadata/iterate_metadata_native.ts
import ts from "typescript";

import { MetadataNative } from "../../../schemas/metadata/MetadataNative";

import { ArrayUtil } from "../../../utils/ArrayUtil";

import { TypeFactory } from "../../TypeFactory";
import { IMetadataIteratorProps } from "./IMetadataIteratorProps";

export const iterate_metadata_native = (
  props: IMetadataIteratorProps,
): boolean => {
  const name: string = getNativeName(
    TypeFactory.getFullName({
      checker: props.checker,
      type: props.type,
      symbol: props.type.getSymbol(),
    }),
  );
  const simple: IClassInfo | undefined = SIMPLES.get(name);
  if (
    simple !== undefined &&
    validate({
      checker: props.checker,
      type: props.type,
      info: simple,
    })
  ) {
    ArrayUtil.take(
      props.metadata.natives,
      (native) => native.name === name,
      () =>
        MetadataNative.create({
          name,
          tags: [],
        }),
    );
    return true;
  }

  for (const generic of GENERICS)
    if (
      name.substring(0, generic.name.length) === generic.name &&
      validate({
        checker: props.checker,
        type: props.type,
        info: generic,
      })
    ) {
      ArrayUtil.take(
        props.metadata.natives,
        (native) => native.name === name,
        () =>
          MetadataNative.create({
            name: generic.name ?? name,
            tags: [],
          }),
      );
      return true;
    }
  return false;
};

const validate = (props: {
  checker: ts.TypeChecker;
  type: ts.Type;
  info: IClassInfo;
}) =>
  (props.info.methods ?? []).every((method) => {
    const returnType = TypeFactory.getReturnTypeOfClassMethod({
      checker: props.checker,
      class: props.type,
      function: method.name,
    });
    return (
      returnType !== null &&
      props.checker.typeToString(returnType).split("<")?.[0] === method.return
    );
  }) &&
  (props.info.properties ?? []).every((property) => {
    const prop = props.checker.getPropertyOfType(props.type, property.name);
    const propType = prop?.valueDeclaration
      ? props.checker.getTypeAtLocation(prop?.valueDeclaration)
      : undefined;
    return (
      propType !== undefined &&
      props.checker.typeToString(propType).split("<")?.[0] === property.type
    );
  });

const getBinaryProps = (className: string): IClassInfo => ({
  name: className,
  methods: [
    ...["indexOf", "lastIndexOf"].map((name) => ({
      name,
      return: "number",
    })),
    ...["some", "every"].map((name) => ({
      name,
      return: "boolean",
    })),
    ...["join", "toLocaleString"].map((name) => ({
      name,
      return: "string",
    })),
    ...["reverse", "slice", "subarray"].map((name) => ({
      name,
      return: className,
    })),
  ],
  properties: ["BYTES_PER_ELEMENT", "length", "byteLength", "byteOffset"].map(
    (name) => ({
      name,
      type: "number",
    }),
  ),
});

const getNativeName = (name: string): string => {
  name = name.split("<")?.[0] ?? "";
  if (name.startsWith('"') && name.slice(0).includes('".'))
    name = name.slice(name.indexOf('".', 1) + 2);
  return name;
};

const SIMPLES: Map<string, IClassInfo> = new Map([
  [
    "Date",
    {
      methods: ["getTime", "getFullYear", "getMonth", "getMinutes"].map(
        (name) => ({
          name,
          return: "number",
        }),
      ),
    },
  ],
  [
    "Boolean",
    {
      methods: [
        {
          name: "valueOf",
          return: "boolean",
        },
      ],
    },
  ],
  [
    "Number",
    {
      methods: [
        ...["toFixed", "toExponential", "toPrecision"].map((name) => ({
          name,
          return: "string",
        })),
        { name: "valueOf", return: "number" },
      ],
    },
  ],
  [
    "String",
    {
      methods: [
        "charAt",
        "concat",
        "valueOf",
        "trim",
        "replace",
        "substring",
      ].map((name) => ({ name, return: "string" })),
    },
  ],
  ...[
    "Uint8Array",
    "Uint8ClampedArray",
    "Uint16Array",
    "Uint32Array",
    "BigUint64Array",
    "Int8Array",
    "Int16Array",
    "Int32Array",
    "BigInt64Array",
    "Float32Array",
    "Float64Array",
  ].map((name) => [name, getBinaryProps(name)] as const),
  ...["ArrayBuffer", "SharedArrayBuffer"].map((className) => {
    const info: IClassInfo = {
      methods: [{ name: "slice", return: className }],
      properties: [{ name: "byteLength", type: "number" }],
    };
    return [className, info] as const;
  }),
  ...["Blob", "File"].map(
    (className) =>
      [
        className,
        {
          methods: [
            { name: "arrayBuffer", return: "Promise" },
            { name: "slice", return: "Blob" },
            { name: "text", return: "Promise" },
          ],
          properties: [
            { name: "size", type: "number" },
            { name: "type", type: "string" },
          ],
        },
      ] satisfies [string, IClassInfo],
  ),
  [
    "DataView",
    {
      methods: [
        "getFloat32",
        "getFloat64",
        "getInt8",
        "getInt16",
        "getInt32",
        "getUint8",
        "getUint16",
        "getUint32",
      ].map((name) => ({
        name,
        return: "number",
      })),
    },
  ],
  [
    "RegExp",
    {
      methods: [
        {
          name: "test",
          return: "boolean",
        },
      ],
    },
  ],
]);
const GENERICS: Array<IClassInfo & { name: string }> = [
  "WeakMap",
  "WeakSet",
].map((name) => ({
  name,
  methods: ["has", "delete"].map((name) => ({
    name,
    return: "boolean",
  })),
}));

interface IClassInfo {
  name?: string;
  methods?: IMethod[];
  properties?: IProperty[];
}
interface IProperty {
  name: string;
  type: string;
}
interface IMethod {
  name: string;
  return: string;
}

File: packages/typia/native/core/factories/internal/metadata/iterate_metadata_native.go
package metadata

import (
  "strings"

  nativeast "github.com/microsoft/typescript-go/shim/ast"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

func Iterate_metadata_native(props IMetadataIteratorProps) bool {
  var symbol *nativeast.Symbol
  name := ""
  if props.Type != nil {
    symbol = props.Type.Symbol()
    name = metadata_type_symbol_base_name(props.Type)
  }
  if name == "" && props.Checker != nil && props.Type != nil {
    apparent := props.Checker.GetApparentType(props.Type)
    if apparent != nil {
      symbol = apparent.Symbol()
    }
    name = metadata_type_symbol_base_name(apparent)
  }
  name = iterate_metadata_native_getNativeName(name)
  if _, ok := iterate_metadata_native_simples[name]; ok {
    // Match the built-in only when the checker resolves this exact symbol from
    // the global type table *and* a runtime authority provides its constructor.
    // A colliding package declaration then falls through to its structural
    // members regardless of whether it was authored in `.ts` or published
    // through `.d.ts` (#2239), and so does a purely user-authored global that
    // merely wins the same global lookup (#2200). Node's authoritative Blob and
    // File module exports are distinct symbols for the same runtime globals, so
    // admit only their value-bearing exact core-module declaration shape as the
    // second identity path (#1568).
    if !metadata_symbol_is_runtime_native_type(props.Checker, symbol, name) {
      return false
    }
    iterate_metadata_native_take(props.Metadata, name)
    return true
  }
  for _, generic := range iterate_metadata_native_generics {
    if name == generic.Name || strings.HasPrefix(name, generic.Name+"<") {
      // Use the same runtime-provenance identity gate for WeakMap/WeakSet. The
      // `+"<"` arity guard (#2181) and generic metadata remain unchanged.
      if !metadata_symbol_is_runtime_global_type(props.Checker, symbol, generic.Name) {
        return false
      }
      iterate_metadata_native_take(props.Metadata, generic.Name)
      return true
    }
  }
  return false
}

func iterate_metadata_native_take(metadata *schemametadata.MetadataSchema, name string) {
  for _, native := range metadata.Natives {
    if native.Name == name {
      return
    }
  }
  metadata.Natives = append(metadata.Natives, schemametadata.MetadataNative_create(schemametadata.MetadataNative{
    Name: name,
    Tags: [][]schemametadata.IMetadataTypeTag{},
  }))
}

func iterate_metadata_native_getNativeName(name string) string {
  if index := strings.Index(name, "<"); index != -1 {
    name = name[:index]
  }
  // `global.` is the qualifier `metadata_symbol_name` produces for a symbol
  // declared inside a `declare global` block, which is how @types/node declares
  // the bare `Blob` and `File` globals. A project that includes the DOM library
  // never showed it, because the DOM declares those names at file scope and the
  // DOM declaration won the name; drop the DOM library — an ordinary Node
  // service compiled with `"lib": ["ES2022"]` — and the only declaration left is
  // the augmented one, so the name arrives as `global.Blob` and this lookup
  // misses before any identity question is asked. Stripping the qualifier only
  // restores the question; whether the declaration is a runtime-provided native
  // or a same-named user global stays the identity gate's answer to give.
  //
  // An ambient module's qualifier is deliberately not stripped here: a lookalike
  // published as `declare module "user-buffer-lookalike"` arrives as
  // `user-buffer-lookalike.Blob` and must keep failing this lookup.
  for _, module := range []string{"node:buffer.", "buffer.", "global."} {
    if strings.HasPrefix(name, module) {
      return strings.TrimPrefix(name, module)
    }
  }
  if strings.HasPrefix(name, "\"") && strings.Contains(name, "\".") {
    if index := strings.Index(name[1:], "\"."); index != -1 {
      name = name[index+3:]
    }
  }
  return name
}

type iterate_metadata_native_classInfo struct {
  Name       string
  Methods    []iterate_metadata_native_method
  Properties []iterate_metadata_native_property
}

type iterate_metadata_native_method struct {
  Name   string
  Return string
}

type iterate_metadata_native_property struct {
  Name string
  Type string
}

var iterate_metadata_native_simples = map[string]iterate_metadata_native_classInfo{
  "Date":              {Name: "Date"},
  "Boolean":           {Name: "Boolean"},
  "BigInt":            {Name: "BigInt"},
  "Number":            {Name: "Number"},
  "String":            {Name: "String"},
  "Uint8Array":        iterate_metadata_native_getBinaryProps("Uint8Array"),
  "Uint8ClampedArray": iterate_metadata_native_getBinaryProps("Uint8ClampedArray"),
  "Uint16Array":       iterate_metadata_native_getBinaryProps("Uint16Array"),
  "Uint32Array":       iterate_metadata_native_getBinaryProps("Uint32Array"),
  "BigUint64Array":    iterate_metadata_native_getBinaryProps("BigUint64Array"),
  "Int8Array":         iterate_metadata_native_getBinaryProps("Int8Array"),
  "Int16Array":        iterate_metadata_native_getBinaryProps("Int16Array"),
  "Int32Array":        iterate_metadata_native_getBinaryProps("Int32Array"),
  "BigInt64Array":     iterate_metadata_native_getBinaryProps("BigInt64Array"),
  "Float32Array":      iterate_metadata_native_getBinaryProps("Float32Array"),
  "Float64Array":      iterate_metadata_native_getBinaryProps("Float64Array"),
  "ArrayBuffer":       {Name: "ArrayBuffer"},
  "SharedArrayBuffer": {Name: "SharedArrayBuffer"},
  "Blob":              {Name: "Blob"},
  "File":              {Name: "File"},
  "DataView":          {Name: "DataView"},
  "RegExp":            {Name: "RegExp"},
}

var iterate_metadata_native_generics = []iterate_metadata_native_classInfo{
  {Name: "WeakMap"},
  {Name: "WeakSet"},
}

func iterate_metadata_native_getBinaryProps(className string) iterate_metadata_native_classInfo {
  return iterate_metadata_native_classInfo{
    Name: className,
    Methods: []iterate_metadata_native_method{
      {Name: "indexOf", Return: "number"},
      {Name: "lastIndexOf", Return: "number"},
      {Name: "some", Return: "boolean"},
      {Name: "every", Return: "boolean"},
      {Name: "join", Return: "string"},
      {Name: "toLocaleString", Return: "string"},
      {Name: "reverse", Return: className},
      {Name: "slice", Return: className},
      {Name: "subarray", Return: className},
    },
    Properties: []iterate_metadata_native_property{
      {Name: "BYTES_PER_ELEMENT", Type: "number"},
      {Name: "length", Type: "number"},
      {Name: "byteLength", Type: "number"},
      {Name: "byteOffset", Type: "number"},
    },
  }
}
